Source/WebKit/Platform/IPC/IPCSemaphore.h

```diff
@@ -85,4 +94,93 @@ class Semaphore {
 #endif
 };
 
+// FIXME: Write proper interruptible implementations for non-COCOA platforms.
+class InterruptibleBinarySemaphoreSignaler {
+    WTF_MAKE_FAST_ALLOCATED;
+    WTF_MAKE_NONCOPYABLE(InterruptibleBinarySemaphoreSignaler);
+public:
+    InterruptibleBinarySemaphoreSignaler(InterruptibleBinarySemaphoreSignaler&& other) = default;
+    InterruptibleBinarySemaphoreSignaler& operator=(InterruptibleBinarySemaphoreSignaler&& other) = default;
+
+    void signal();
+
+    void encode(Encoder&) &&;
+    static std::optional<InterruptibleBinarySemaphoreSignaler> decode(Decoder&);
+
+private:
+    friend struct InterruptibleBinarySemaphore;
+#if PLATFORM(COCOA)
+    InterruptibleBinarySemaphoreSignaler(MachSendRight&& sendRight)
+        : m_sendRight(WTFMove(sendRight))
+    { }
+#else
+    InterruptibleBinarySemaphoreSignaler(Semaphore&& semaphore)
+        : m_semaphore(WTFMove(semaphore))
+    { }
+#endif
+
+#if PLATFORM(COCOA)
+    MachSendRight m_sendRight;
+#else
+    Semaphore m_semaphore;
+#endif
+};
+
+class InterruptibleBinarySemaphoreWaiter {
+    WTF_MAKE_FAST_ALLOCATED;
+    WTF_MAKE_NONCOPYABLE(InterruptibleBinarySemaphoreWaiter);
+public:
+#if PLATFORM(COCOA)
+    ~InterruptibleBinarySemaphoreWaiter();
+    InterruptibleBinarySemaphoreWaiter(InterruptibleBinarySemaphoreWaiter&& other)
+        : m_receiveRight(WTFMove(other.m_receiveRight))
+    {
+        other.m_receiveRight = MACH_PORT_NULL;
+    }
+
+    InterruptibleBinarySemaphoreWaiter& operator=(InterruptibleBinarySemaphoreWaiter&& other)
+    {
+        m_receiveRight = WTFMove(other.m_receiveRight);
+        other.m_receiveRight = MACH_PORT_NULL;
+        return *this;
+    }
+#else
+    InterruptibleBinarySemaphoreWaiter(InterruptibleBinarySemaphoreWaiter&& other) = default;
+    InterruptibleBinarySemaphoreWaiter& operator=(InterruptibleBinarySemaphoreWaiter&& other) = default;
+#endif
+
+    bool wait();
+    bool waitFor(Timeout);
+
+private:
+    friend struct InterruptibleBinarySemaphore;
+#if PLATFORM(COCOA)
+    InterruptibleBinarySemaphoreWaiter(mach_port_t port)
+        : m_receiveRight(port)
+    { }
+#else
+    InterruptibleBinarySemaphoreWaiter(Semaphore&& semaphore)
+        : m_semaphore(WTFMove(semaphore))
+    { }
+#endif
+
+#if PLATFORM(COCOA)
+    Lock m_lock;
+    mach_port_t m_receiveRight;
```

Comment:
WTF_GUARDED_BY(m_lock); perhaps?

Reply:
In fact, the lock gives the impression that this would be thread-safe, but I'm not sure that makes sense.
For example, the move operation is not.
Contrast with IPC::Semaphore, there's no lock there.